crates/control/src/game_state_filter.rs

```diff
@@ -116,23 +116,46 @@ fn next_filtered_state(
     ball_detected_far_from_any_goal: bool,
 ) -> State {
     match (current_state, game_controller_state.game_state) {
-        (State::Initial | State::Ready | State::Finished, _)
-        | (
-            State::Set,
-            GameState::Initial | GameState::Ready | GameState::Playing | GameState::Finished,
-        )
+        (State::Finished, _) => State::Finished,
+
+        (
+            State::TentativeFinished {
+                time_when_finished_clicked,
+            },
+            GameState::Finished,
+        ) if cycle_start_time
+            .duration_since(time_when_finished_clicked)
+            .unwrap()
+            .as_secs_f32()
+            >= 4.0 =>
+        {
+            State::Finished
+        }
+        (State::TentativeFinished { .. }, game_state) if game_state != GameState::Finished => {
+            State::from_game_state(game_state)
+        }
+        (
+            State::TentativeFinished {
+                time_when_finished_clicked,
+            },
+            _,
+        ) => State::TentativeFinished {
+            time_when_finished_clicked,
+        },
```

Comment:
Can you move this up as `else` case for `if` in line 126?

Reply:
Done

Reply:
Nice 👍🏻